### ISSUE TITLE
[Parity 16/16] Graph compile parity - codegen, guards/recompile, dynamic shapes, AOTAutograd parity, CUDA Graph training

### ISSUE THREAD

**Issue by ooples** [ooples/AiDotNet.Tensors#225]
Part of epic #209.

## Summary

`torch.compile` (TorchInductor + Dynamo + AOTAutograd) is PyTorch's flagship performance story. Our graph infrastructure — `LazyTensorScope`, `CompiledInferencePlan`, `CompiledTrainingPlan`, `AutoTracer`, `SymbolicShape`, and a rich `Engines/Optimization/` pass library — is already substantial and in several dimensions *ahead* of Inductor. What's missing to beat PyTorch head-to-head: **custom-kernel codegen**, **guard/recompilation** correctness, **dynamic-shape completeness**, **AOTAutograd parity** for training compile, and **CUDA Graph capture of training** (we have inference).

### Capture (Dynamo)
- Bytecode-level Python capture with full control-flow rewriting
- Recompilation guards on tensor shape, dtype, device, stride, memory-format
- `TORCH_LOGS=recompiles,guards` visibility

### AOTAutograd
- Captures forward + backward as a joint graph
- Runs graph-level dead-code elimination + in-place-op reinsertion
- Min-cut partitioner splits forward/backward to minimize activation retention

### TorchInductor (codegen)
- Triton kernels emitted per fusion
- C++/OpenMP for CPU
- Template matmul
- Horizontal fusion for reductions
- Loop reordering, tiling, vectorization

### Dynamic shapes
- `torch.compile(dynamic=True)` — symbolic sizes
- Constraint propagation + simplification
- Size hints + recompilation budget

### CUDA Graph integration
- `torch.cuda.make_graphed_callables` — capture forward + backward + optimizer

## Current state in the repo

- `Engines/Compilation/*`: `LazyTensorScope`, `CompiledInferencePlan`, `CompiledTrainingPlan`, `AutoTracer`, `AutoTrainingCompiler`, `SymbolicShape`.
- `Engines/Optimization/*`: ~20 passes including several *unique* to us (`SpectralDecompositionPass`, `TensorCodecOptimizer`, `DataflowFusionPass`, `AttentionFusionPass`, `TileSchedulingPass`).
- `Engines/Autodiff/SymbolicBackwardGraph.cs`, `CompiledBackwardGraph.cs`, `BackwardCSEPass.cs`, `BackwardPruningPass.cs`, `AlgebraicSimplificationPass.cs`.
- `Engines/Gpu/CudaGraphScope.cs` — inference graph capture exists.

Missing: custom kernel codegen (we always compose pre-built kernels), guard/recompile semantics, full dynamic-shape coverage, AOTAutograd-equivalent joint-graph partitioning for training, CUDA Graph of training loop.

## How we beat PyTorch

1. **Keep our pass moat, extend it**: Spectral + Codec + Dataflow + Attention fusion aren't in Inductor. Every codegen target we add runs through this pipeline first. So our codegen starts with a simplified graph Inductor never sees.
2. **Codegen to multiple backends, not just Triton**: emit Triton (CUDA), HIP (ROCm), MSL (Metal), GLSL compute (Vulkan), WGSL (WebGPU), AVX-512 C# intrinsics (CPU). PyTorch = Triton + C++/OpenMP only.
3. **Codegen *through* our NCHWc layout**: Inductor doesn't know NCHWc; we produce kernels that stay in packed layout. Already ahead; codify it in the codegen contract.
4. **Sub-byte codegen**: emit kernels for int1/int2/int3/int4/NF4/FP4/FP8. Unique.
5. **Guard system backed by `SymbolicShape`**: guards are cheap (bit-compare of a packed shape tuple) and recompile decisions use `AutotuneCache` — we recompile only when the cache says the new shape bucket is large enough to benefit. PyTorch recompiles aggressively and tax-es with guard eval cost.
6. **Joint-graph min-cut for AOTAutograd equivalent**: re-use `BackwardCSEPass` + `BackwardPruningPass` + new `ActivationCheckpointPartitioner` that solves a min-cut problem. Already have pieces — assemble.
7. **CUDA Graph capture of training loop**: `MixedPrecisionTrainingLoop` (Wave 1) + `CudaGraphScope` + fused optimizer = full-step graph capture. Requires deterministic allocators (we have `TensorArena`) and graph-safe RNG (from GPU-primitives issue). Target end-to-end training step replay that's 20–30% faster than eager.
8. **Dynamic shapes that don't recompile**: our tile scheduler already handles variable M/N/K in GEMM via autotune buckets. Extend to convs + reductions. Target: ≤2 recompiles across a batch-size sweep of 8/16/32/64/128.
9. **`.NET native codegen` via `System.Reflection.Emit` / source generators**: generate .NET methods that call our kernels with zero virtual dispatch. PyTorch's CPU fallback is slow C++ because Python glue taxes each call.

### Custom kernel codegen
- [ ] IR definition for fused pointwise / reduction / matmul / attention
- [ ] Triton emitter (CUDA)
- [ ] HIP emitter
- [ ] MSL emitter (Metal)
- [ ] WGSL emitter (WebGPU)
- [ ] GLSL compute emitter (Vulkan)
- [ ] AVX-512 C# intrinsic emitter (CPU)
- [ ] Sub-byte codegen paths

### Dynamic shapes
- [ ] Audit `SymbolicShape.cs` coverage
- [ ] Constraint propagation across fusion passes
- [ ] Recompile-budget + shape-bucketing based on `AutotuneCache`
- [ ] Shape guards with sub-microsecond check
- [ ] Logging equivalent to `TORCH_LOGS=recompiles,guards`

### AOTAutograd equivalent
- [ ] Joint forward+backward graph capture (have pieces in `CompiledTrainingPlan`)
- [ ] Min-cut activation partitioner
- [ ] Dead-code elimination across joint graph
- [ ] In-place-op reinsertion post-optimization

### CUDA Graph capture of training
- [ ] Deterministic allocator state via `TensorArena` + replay-aware `GpuWorkspace`
- [ ] Graph-safe RNG integration
- [ ] `MakeGraphedTrainingStep(model, loss, optimizer)` utility
- [ ] Capture → replay → correctness test

### Observability
- [ ] Per-pass timing (ties to profiler issue)
- [ ] Recompile log with reason strings
- [ ] Autotune cache hit/miss stats

## Acceptance criteria

- Training-mode compile correctness on a small model (MLP / small ViT / LSTM) — loss curve matches eager within tolerance.
- Benchmark: compiled training step ≥ 20% faster than our eager path on a standard ResNet-50 benchmark; competitive with `torch.compile` fullgraph on the same model.
- Dynamic batch-size sweep 8→128 recompiles ≤ 2 times.
- CUDA Graph of training loop: correct over N iterations, ≥20% faster than non-graphed.
- Custom codegen outputs a Triton kernel for a fused `x * sigmoid(x) + bias` on CUDA; correctness tested; faster than our pre-built generic kernel.
- Min-cut partitioner demonstrates lower peak memory than the naive checkpoint strategy.

## Implementation notes

- IR under `Engines/Compilation/Codegen/Ir/`.
- Emitters under `Engines/Compilation/Codegen/{Triton,Hip,Msl,Wgsl,Glsl,AvxCs}/`.
- Joint graph partitioner alongside `SymbolicBackwardGraph.cs`.
- CUDA Graph training capture under `Engines/Gpu/CudaGraphTrainingScope.cs`.
- Guard system alongside `CompiledModelCache.cs`.

**Comment by ooples** [comment]
# Implementation plan — detailed

## 1. Current state

**Already ahead in several dimensions:**
- `Engines/Compilation/*`: `LazyTensorScope`, `CompiledInferencePlan`, `CompiledTrainingPlan`, `AutoTracer`, `AutoTrainingCompiler`, `SymbolicShape`.
- `Engines/Optimization/*`: ~20 passes incl. unique `SpectralDecompositionPass`, `TensorCodecOptimizer`, `DataflowFusionPass`, `AttentionFusionPass`, `TileSchedulingPass`.
- `Engines/Autodiff/SymbolicBackwardGraph.cs`, `CompiledBackwardGraph.cs`, `BackwardCSEPass.cs`, `BackwardPruningPass.cs`, `AlgebraicSimplificationPass.cs`.
- `Engines/Gpu/CudaGraphScope.cs` — inference graph capture ✓.

**Missing:** custom kernel codegen, guard/recompile semantics, dynamic-shape completeness, AOTAutograd-equivalent joint-graph partitioning for training, CUDA Graph of training loop.

## 2. Scope — full inventory

### Custom kernel codegen
```
IR definition (Engines/Compilation/Codegen/Ir/)
  - FusedPointwise, FusedReduction, FusedMatMul, FusedAttention nodes
Triton emitter (CUDA)
HIP emitter (ROCm)
MSL emitter (Metal)
WGSL emitter (WebGPU)
GLSL compute emitter (Vulkan)
AVX-512 C# intrinsic emitter (CPU; via System.Reflection.Emit + source gen)
Sub-byte codegen paths (int1/2/3/4/NF4/FP4/FP8)
```

### Dynamic shapes
```
Audit SymbolicShape.cs coverage
Constraint propagation across fusion passes
Recompile-budget + shape-bucketing based on AutotuneCache
Shape guards with sub-microsecond check (packed-tuple bit compare)
TORCH_LOGS=recompiles,guards equivalent: LogRecompiles, LogGuards
```

### AOTAutograd equivalent
```
Joint forward+backward graph capture
Min-cut activation partitioner
Dead-code elimination across joint graph
In-place-op reinsertion post-optimization
```

### CUDA Graph capture of training
```
Deterministic allocator state via TensorArena + replay-aware GpuWorkspace
Graph-safe RNG integration (Philox from #213/#219)
MakeGraphedTrainingStep(model, loss, optimizer) utility
Capture → replay → correctness test
```

### Observability
```
Per-pass timing (ties to #220 profiler)
Recompile log with reason strings
Autotune cache hit/miss stats
```

**Total: ~40 surfaces across codegen + dynamic-shape + AOTAutograd + CUDAGraph + observability.**

## 3. Win vectors

1. **Pass moat extended**: Spectral + Codec + Dataflow + Attention fusion aren't in Inductor. Every codegen target starts with a simplified graph Inductor never sees.
2. **Codegen to 6 backends, not 2**: Triton (CUDA) + HIP + MSL + WGSL + GLSL + AVX-512 C#. PyTorch = Triton + C++/OpenMP.
3. **Codegen *through* NCHWc layout**: Inductor doesn't know NCHWc; we stay packed end-to-end.
4. **Sub-byte codegen**: emit kernels for int1/2/3/4/NF4/FP4/FP8 — unique.
5. **Guard system backed by `SymbolicShape`**: packed-tuple bit-compare (sub-microsecond). Recompile decisions use `AutotuneCache`. PyTorch recompiles aggressively.
6. **Joint-graph min-cut for AOTAutograd equivalent**: reuse `BackwardCSEPass` + `BackwardPruningPass` + new `ActivationCheckpointPartitioner` solving min-cut.
7. **CUDA Graph capture of training loop**: existing `CudaGraphScope` + deterministic `TensorArena` + fused optimizer + Philox RNG = full-step graph capture. Target 20–30% over eager.
8. **Dynamic shapes that don't recompile**: tile scheduler handles variable M/N/K via autotune buckets. Target ≤2 recompiles across batch-size sweep 8/16/32/64/128.
9. **.NET native codegen via `System.Reflection.Emit` / source generators**: methods calling kernels with zero virtual dispatch.

## 4. Per-axis beat targets

| Axis | Target |
|---|---|
| Training compile correctness | loss curve matches eager within tolerance |
| Compiled ResNet-50 training step | ≥20% faster than our eager; ≥ parity with `torch.compile` fullgraph |
| Dynamic batch-size sweep 8→128 recompiles | ≤2 |
| CUDA Graph of training loop | ≥20% over non-graphed |
| Custom codegen (fused `x * sigmoid(x) + bias`) | correctness + ≥ our pre-built generic |
| Min-cut partitioner | lower peak memory than naive checkpoint |
| 6-backend codegen | all targets emit + execute canonical fused pointwise kernel |
| Guard eval cost | sub-microsecond |
| Recompile logging | equivalent to `TORCH_LOGS=recompiles,guards` |

## 5. File inventory

**New (~100):**
- `Engines/Compilation/Codegen/Ir/` — IR definitions + utility passes (~15)
- `Engines/Compilation/Codegen/Triton/` — CUDA emitter (~10)
- `Engines/Compilation/Codegen/Hip/` (~10)
- `Engines/Compilation/Codegen/Msl/` (~10)
- `Engines/Compilation/Codegen/Wgsl/` (~10)
- `Engines/Compilation/Codegen/Glsl/` (~10)
- `Engines/Compilation/Codegen/AvxCs/` (~10)
- `Engines/Compilation/Codegen/SubByte/` (~8)
- `Engines/Compilation/Joint/` — JointGraphPartitioner, MinCut, DceJointGraph, InPlaceReinsertion (~12)
- `Engines/Gpu/CudaGraphTrainingScope.cs` + replay-aware allocator (~5)

**Modified (~15):** `SymbolicShape.cs`, `CompiledTrainingPlan.cs`, `LazyTensorScope.cs`, `LazyGraphCompiler.cs`, `CompiledModelCache.cs` (guards), `CudaGraphScope.cs`, `TensorArena.cs` (replay-aware), `GpuWorkspace.cs`, `SymbolicBackwardGraph.cs` (joint-graph hook), `BackwardCSEPass.cs`, `BackwardPruningPass.cs`, `AlgebraicSimplificationPass.cs`.

**Tests (~30):** codegen correctness per backend (canonical fused kernel), dynamic-shape recompile budget, AOTAutograd joint-graph min-cut parity, CUDA Graph training replay.

**Benchmarks (~6):** `TrainingCompileBenchmarks`, `DynamicShapeSweepBenchmarks`, `CudaGraphTrainingBenchmarks`, `CodegenPerBackendBenchmarks`, `MinCutPartitionerBenchmarks`, `GuardEvalBenchmarks`.

## 6. Acceptance gate

1. Compiled training correctness on MLP / small ViT / LSTM (loss matches eager).
2. Compiled training step ≥20% faster than eager on ResNet-50; competitive with `torch.compile` fullgraph.
3. Dynamic batch-size sweep recompiles ≤2.
4. CUDA Graph of training loop correct over N iterations; ≥20% faster.
5. Custom codegen emits correct Triton kernel for fused pointwise; faster than pre-built generic.
6. Min-cut partitioner lower peak memory than naive.
7. 6 codegen backends each emit and execute canonical fused pointwise kernel.
8. Recompile/guard logs comparable to PyTorch's.

## 7. Timeline — ~12 weeks

W1 IR definition + utility passes.
W2 Triton emitter (CUDA) + correctness test.
W3 HIP emitter + MSL emitter.
W4 WGSL + GLSL emitters.
W5 AVX-512 C# emitter via source gen / Reflection.Emit.
W6 Sub-byte codegen paths.
W7 Guard system via SymbolicShape + packed bit-compare.
W8 Dynamic shape + recompile budget + logging.
W9 Joint-graph capture + min-cut partitioner.
W10 DCE + in-place reinsertion + AOTAutograd parity tests.
W11 CUDA Graph training capture + deterministic allocator.
W12 Benchmarks + PR.

## 8. Dependencies

- **Blocks**: nothing (final issue — builds on everything else).
- **Blocked by**: #214 (vmap + forward-mode to feed joint-graph), #210 (dynamic-shape ops for sort/gather), #219 (NVTX auto-emit for observability; cuRAND for graph-safe RNG), #213 (Philox for deterministic RNG in captured graph), #224 (CUDA Graph-safe optimizers).

## 9. Risks

- **Custom codegen correctness across 6 backends** — highest test burden; per-backend canonical kernel suite mandatory.
- **AOTAutograd min-cut partitioner** — NP-hard in general; use approximate solver + empirical validation.
- **CUDA Graph training replay determinism** — allocator + RNG + order must all be deterministic; co-develop with #213, #219, #224.
- **6 codegen backends = huge maintenance surface** — document the "canonical kernel" contract so each backend only handles a defined subset well.

---

*Part of epic #209. Plan ratified 2026-04-19.*
